[10.x] Replace Deprecated DBAL Comparator creation with schema aware Comparator

liamh101 opened the pull request:
Reference to change in DBAL and why this needs to be done: https://github.com/doctrine/dbal/pull/4746

This is currently causing this bug in Laravel where certain migration changes are ignored: https://github.com/laravel/framework/issues/46492

Not sure if this is affecting other database types.

An example project of the existing bug can be found here: https://github.com/liamh101/laravel-column-bug

liamh101 committed:
Replace Deprecated DBAL Comparator creation with schema aware Comparator

Reference to change in DBAL: https://github.com/doctrine/dbal/pull/4746

liamh101 committed:
Add additional test for changing text columns to different text columns

liamh101 committed:
Revert test change

liamh101 committed:
Remove redundant use statement

hafezdivandari commented:
@liamh101 What about `TableDiff` on this function, do we have to replace that too?

https://github.com/laravel/framework/blob/7d15f7eef442633cff108f83d9fe43d8c3b8b76f/src/Illuminate/Database/Schema/Grammars/Grammar.php#L324-L331

liamh101 commented:
@hafezdivandari Just need to do the `introspectTable` before creating the `TableDiff` class, and I believe the `tap` function call is now redundant. I'll get that changed and see if there are any side effects.

hafezdivandari commented:
@liamh101 isn't it better to change this function and use it on `ChangeColumn` class too if possible?

liamh101 committed:
Rework GetDoctrineTableDiff to remove deprication call due to missing constructor parameter

liamh101 committed:
Change Doctrine Integration tests removing Binary additions

These are no longer included as provider in now schema aware when generating the SQL

liamh101 commented:
@hafezdivandari can I do some further cleaning as a follow-up PR? I don't want to change too much and introduce additional bugs.

My feeling is that this bug can go under the radar on many projects and have massive consequences, partially multi-tenant projects where fresh database migrations can happen regularly.

hafezdivandari commented:
@liamh101 sure, it was just a suggestion. But it seems that `RenameColumn` class has the same issue, I think we should not use `TableDiff` directly and use `$schema->createComparator()->compareTables()`. Does renaming a `mediumText` using DBAL column works right now?

liamh101 committed:
Add additional possibilities for output SQL

liamh101 committed:
Adding missing comma

liamh101 committed:
Add missing Binary in test

liamh101 commented:
Renaming a column will work. However, changing a text type using the default Comparator won't work.

driesvints reviewed (approved):
Thanks for digging more into this @liamh101. I feel like we should give this a shot.

liamh101 commented:
Sure!

- Replace deprecated DBAL Comparator creation for a Schema Aware Comparator. ( `ChangeColumn.php`) Reasoning why DBAL deprecated this can be found here:  https://github.com/doctrine/dbal/pull/4746
- Clean up the additional deprecated function call to DBAL in `Grammar.php` however does not affect any behaviour.
- Changes in `DatabaseSchemaBlueprintIntegrationTest.php` to account for SQL changes present between stable and lowest versions of `DBAL 3.x` .
- Add schema builder test to validate text types change in MYSQL. The current version of Laravel is bugged with existing and new migrations. A project demonstrating this issue can be found here: https://github.com/liamh101/laravel-column-bug

No breaking changes were introduced in this PR, just a bug fix.

Once this has been reviewed and merged, I will investigate other places Laravel might use deprecated DBAL functionality and replace them. This PR is to fix the migration issues introduced when upgrading DBAL from 2.x to 3.x.

taylorotwell committed:
Update Grammar.php

taylorotwell commented:
Thanks
